plugins/steelseries/fu-steelseries-sonic.c

```diff
@@ -0,0 +1,1238 @@
+	guint8 iface_idx;
+	guint8 ep;
+	gsize in_size;
+};
+
+G_DEFINE_TYPE(FuSteelseriesSonic, fu_steelseries_sonic, FU_TYPE_USB_DEVICE)
+
+static gboolean
+fu_steelseries_sonic_command(FuDevice *device, guint8 *data, gboolean answer, GError **error)
+{
+	FuSteelseriesSonic *self = FU_STEELSERIES_SONIC(device);
+	GUsbDevice *usb_device = fu_usb_device_get_dev(FU_USB_DEVICE(device));
+	gsize actual_len = 0;
+	gboolean ret;
+
+	ret = g_usb_device_control_transfer(usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_CLASS,
+					    G_USB_DEVICE_RECIPIENT_INTERFACE,
+					    0x09,
+					    0x0200,
+					    self->iface_idx,
+					    data,
+					    STEELSERIES_BUFFER_CONTROL_SIZE,
+					    &actual_len,
+					    STEELSERIES_TRANSACTION_TIMEOUT,
+					    NULL,
+					    error);
+	if (!ret) {
+		g_prefix_error(error, "failed to do control transfer: ");
+		return FALSE;
+	}
+	if (actual_len != STEELSERIES_BUFFER_CONTROL_SIZE) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "only wrote %" G_GSIZE_FORMAT "bytes",
+			    actual_len);
+		return FALSE;
+	}
+
+	/* cleanup the buffer before receiving any data */
+	memset(data, 0x00, STEELSERIES_BUFFER_CONTROL_SIZE);
+
+	/* do not expect the answer from device */
+	if (answer != TRUE)
+		return TRUE;
+
+	ret = g_usb_device_interrupt_transfer(usb_device,
+					      self->ep,
+					      data,
+					      self->in_size,
+					      &actual_len,
+					      STEELSERIES_TRANSACTION_TIMEOUT,
+					      NULL,
+					      error);
+	if (!ret) {
+		g_prefix_error(error, "failed to do EP transfer: ");
+		return FALSE;
+	}
+	if (actual_len != self->in_size) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "only read %" G_GSIZE_FORMAT "bytes",
+			    actual_len);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_wireless_status(FuDevice *device,
+				     SteelseriesSonicWirelessStatus *status,
+				     GError **error)
+{
+	guint8 data[STEELSERIES_BUFFER_CONTROL_SIZE] = {0};
+	const guint16 opcode = 0xE8U; /* dongle */
+	guint8 value;
+
+	if (!fu_common_write_uint8_safe(data, sizeof(data), 0x0, opcode, error))
+		return FALSE;
+
+	if (g_getenv("FWUPD_STEELSERIES_SONIC_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "WirelessStatus", data, sizeof(data));
+	if (!fu_steelseries_sonic_command(device, data, TRUE, error))
+		return FALSE;
+
+	if (g_getenv("FWUPD_STEELSERIES_SONIC_VERBOSE") != NULL)
+		fu_common_dump_raw(G_LOG_DOMAIN, "WirelessStatus", data, sizeof(data));
+	if (!fu_common_read_uint8_safe(data, sizeof(data), 0x0, &value, error))
+		return FALSE;
+	*status = value;
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_steelseries_sonic_battery_state(FuDevice *device, guint16 *value, GError **error)
+{
+	guint8 data[STEELSERIES_BUFFER_CONTROL_SIZE] = {0};
+	const guint16 opcode = 0xAAU;
+	const guint8 bat_mode = 1U; /* percentage */
```

Comment:
the `U` is confusing me, maybe `0x01` instead?

Reply:
Indeed.